bin/node-template/node/src/service.rs

```diff
@@ -24,100 +21,107 @@ native_executor_instance!(
 	node_template_runtime::native_version,
 );
 
-/// Starts a `ServiceBuilder` for a full service.
-///
-/// Use this macro if you don't actually need the full service, but just the builder in order to
-/// be able to perform chain operations.
-macro_rules! new_full_start {
-	($config:expr) => {{
-		use std::sync::Arc;
-		use sp_consensus_aura::sr25519::AuthorityPair as AuraPair;
-
-		let mut import_setup = None;
-		let inherent_data_providers = sp_inherents::InherentDataProviders::new();
-
-		let builder = sc_service::ServiceBuilder::new_full::<
-			node_template_runtime::opaque::Block,
-			node_template_runtime::RuntimeApi,
-			crate::service::Executor
-		>($config)?
-			.with_select_chain(|_config, backend| {
-				Ok(sc_consensus::LongestChain::new(backend.clone()))
-			})?
-			.with_transaction_pool(|builder| {
-				let pool_api = sc_transaction_pool::FullChainApi::new(
-					builder.client().clone(),
-				);
-				Ok(sc_transaction_pool::BasicPool::new(
-					builder.config().transaction_pool.clone(),
-					std::sync::Arc::new(pool_api),
-					builder.prometheus_registry(),
-				))
-			})?
-			.with_import_queue(|
-				_config,
-				client,
-				mut select_chain,
-				_transaction_pool,
-				spawn_task_handle,
-				registry,
-			| {
-				let select_chain = select_chain.take()
-					.ok_or_else(|| sc_service::Error::SelectChainRequired)?;
-
-				let (grandpa_block_import, grandpa_link) = sc_finality_grandpa::block_import(
-					client.clone(),
-					&(client.clone() as Arc<_>),
-					select_chain,
-				)?;
-
-				let aura_block_import = sc_consensus_aura::AuraBlockImport::<_, _, _, AuraPair>::new(
-					grandpa_block_import.clone(), client.clone(),
-				);
-
-				let import_queue = sc_consensus_aura::import_queue::<_, _, _, AuraPair, _>(
-					sc_consensus_aura::slot_duration(&*client)?,
-					aura_block_import,
-					Some(Box::new(grandpa_block_import.clone())),
-					None,
-					client,
-					inherent_data_providers.clone(),
-					spawn_task_handle,
-					registry,
-				)?;
-
-				import_setup = Some((grandpa_block_import, grandpa_link));
-
-				Ok(import_queue)
-			})?;
+type FullClient = sc_service::TFullClient<Block, RuntimeApi, Executor>;
+type FullBackend = sc_service::TFullBackend<Block>;
+type GrandpaBlockImport = sc_finality_grandpa::GrandpaBlockImport<
+	FullBackend, Block, FullClient, SelectChain
+>;
+type SelectChain = sc_consensus::LongestChain<FullBackend, Block>;
+type GrandpaLink = sc_finality_grandpa::LinkHalf<Block, FullClient, SelectChain>;
+type FullPool = sc_transaction_pool::BasicPool<
+	sc_transaction_pool::FullChainApi<FullClient, Block>, Block
+>;
```

Comment:
I'm not a fan of having to use type definitions like this, but I'm not sure what the alternative would be. Having everything in `ServiceParams` be `Box<dyn ...>` perhaps?

Reply:
I'm not sure you can as Box is not mutable. (Otherwise I would say yes it can be good if nothing is blocking it but I don't think it will work)

Reply:
I ended up creating a macro to do this instead.